### PR TITLE
Fixes for substitute stations

### PR DESCRIPTION
**Pull request by juliayyan**
This fixes the errors in model-building from the last meeting.

Changes:
- bugfix so that `np.stage` is not erroneously updated in `expandedstages()` 
- precompute the substitute stations in `np` construction to save time
- slightly more restrictive criteria for being substitute stations to prevent an explosion of route options

### CODE REVIEW

File: src/network/networkproblem.jl [hunk 1]
```diff
@@ -92,7 +93,8 @@ function TransitNetworkProblem{T}(
                     @assert etype == :platform
                     3 # minutes to transfer
                 end
-            end
+            end,
+        maxdist::Float64 = 0.25
```
Comment (yeesian) [inline src/network/networkproblem.jl]: we should indicate in a comment/docstring whether it's in miles or km

File: src/network/networkmethods.jl [hunk 2]
```diff
@@ -220,16 +220,16 @@ function expandedstages(np::TransitNetworkProblem, u::Int, v::Int, maxdist::Floa
     return od, routes
 end 
 
-"computes distance between stn u and stn v"
-function haversinedistance(np::TransitNetworkProblem, u::Int, v::Int)
-    lat1 = np.latlngs[u,1]
-    lat2 = np.latlngs[v,1]
-    lon1 = np.latlngs[u,2]
-    lon2 = np.latlngs[v,2]
+"computes distance between (lat1, lon1) and (lat2, lon2)"
```
Comment (yeesian) [inline src/network/networkmethods.jl]: distance in miles/km?

File: src/network/networkmethods.jl [hunk 3]
```diff
@@ -220,16 +220,16 @@ function expandedstages(np::TransitNetworkProblem, u::Int, v::Int, maxdist::Floa
...
+function haversinedistance(lat1::Float64, lat2::Float64, lon1::Float64, lon2::Float64)
     2 * 3959 * asin(sqrt(
         sind((lat2-lat1)/2)^2 + cosd(lat1) * cosd(lat2) * sind((lon2-lon1)/2)^2
     ))
-end
+end 
+
+"computes distance between stn u and stn v"
```
Comment (yeesian) [inline src/network/networkmethods.jl]: distance in miles/km?

File: src/network/networkproblem.jl [hunk 4]
```diff
@@ -539,14 +541,45 @@ function TransitNetworkProblem{T}(
...
+    # - it is less than mindist away from u
+    # - it is not on the same line as u
+    # - it is not on the same line as any closer station to u
+    # (i) substitutes
+    # -----
+    linked = zeros(Bool, nstns, nstns)
+    for l in 1:length(lines), i in 1:length(lines[l]), j in i:length(lines[l])
+        linked[lines[l][i], lines[l][j]] = true
+        linked[lines[l][j], lines[l][i]] = true
+    end
```
Comment (yeesian) [inline src/network/networkproblem.jl]: I believe the following should work too:
```julia
for line in lines, i in line, j in line
    linked[i, j] = linked[j, i] = true
end
```

File: src/network/networkmethods.jl (hunk 2 above)
Comment (yeesian) [reply src/network/networkmethods.jl]: [Okay, I think it's km here](https://github.com/yeesian/Turf.jl/blob/056df20eff2950a25a9ec2cc1959185b7ce22371/src/measurement.jl#L3-L12)

File: src/network/networkproblem.jl [hunk 5]
```diff
@@ -65,6 +65,7 @@ mutable struct TransitNetworkProblem{T}
     platforms::Dict{Int,Vector{Int}} # [stn => [list of lines]]
     networkplatforms::Vector{Tuple{Int,Int}}
     platformindex::Dict{Tuple{Int,Int},Int}
+    substitutes::Vector{Vector{Int}}
```
Comment (yeesian) [inline src/network/networkproblem.jl]: we should add it to the docstring

**Comment by juliayyan** [comment]
This is nice -- thanks! okay to merge?

**Comment by yeesian** [comment]
Yup!

**Pull request merged**